File: hashFileName.c
#import <string.h>
#include <stdio.h>
#include <stdint.h>

static inline uint16_t rotl16(uint16_t value, uint8_t shift) {
    // Masking the shift with 15 prevents undefined behavior 
    // if shift is >= 16
    return (value << (shift & 15)) | (value >> ((16 - shift) & 15));
}

int hashFileNameMM4( char *name, int len )
{
  int i, h;

  if( len < 1 ) return( -1 );

  for( i = 1, h = name[0] ; i < len ; h += name[i++] )
  {
    // Rotate the bits in 'h' right 7 places
    // In assembly it would be: ror h, 7
    // 01234567 89ABCDEF -> 9ABCDEF0 12345678
    // 0x007F = 00000000 01111111
    // 0xFF80 = 11111111 10000000
    h = ( h & 0x007F ) << 9 | ( h & 0xFF80 ) >> 7;
  }

  return( h );
}

uint16_t hashFileNameMM3(const char* fileName, int len)
{
  uint16_t hash = 0;
  while (0 != *fileName)
  {
    uint8_t c = ((*fileName & 0x7F) < 0x60) ? *fileName : *fileName - 0x20;
    hash = rotl16(hash, 9);  // xchg bl, bh | rol bx, 1
    hash += c;
    fileName++;
  }
  return hash;
}

int main(int argc,char* argv[]){

  if (strlen(argv[1]) >= 0){
    int mm4hash = hashFileNameMM4(argv[1], strlen(argv[1]));
    printf("MM4 Hash: %d\n", mm4hash);
    printf("MM4 Dec : %04x\n", mm4hash);
    int mm3hash = hashFileNameMM3(argv[1], strlen(argv[1]));
    printf("MM3 Hash: %d\n", mm3hash);
    printf("MM3 Dec : %04x\n", mm3hash);
  }

}

File: hashFileName.py
import sys

def rotl16(value, shift):
    """Circularly shift a 16-bit value to the left."""
    # Masking ensures we stay within 16-bit bounds
    shift &= 15
    return ((value << shift) | (value >> (16 - shift))) & 0xFFFF

def hash_file_name_mm4(name):
    length = len(name)
    if length < 1:
        return -1

    # Convert string to list of ordinals for processing
    name_bytes = [ord(c) for c in name]
    h = name_bytes[0]

    for i in range(1, length):
        # Emulate the manual bit rotation from the C code
        # h = ( h & 0x007F ) << 9 | ( h & 0xFF80 ) >> 7;
        part1 = (h & 0x007F) << 9
        part2 = (h & 0xFF80) >> 7
        h = (part1 | part2) & 0xFFFF # Keep it 16-bit
        
        # Add the next character
        h = (h + name_bytes[i]) & 0xFFFF

    return h

def hash_file_name_mm3(name):
    hash_val = 0
    for char in name:
        c_ord = ord(char)
        # Check if character should be converted to uppercase (basic ASCII logic)
        c = c_ord if (c_ord & 0x7F) < 0x60 else c_ord - 0x20
        
        # Rotate left 9 bits
        hash_val = rotl16(hash_val, 9)
        # Add the processed character
        hash_val = (hash_val + c) & 0xFFFF
        
    return hash_val

def main():
    if len(sys.argv) < 2:
        print("Usage: python script.py <string>")
        return

    input_str = sys.argv[1]

    mm4 = hash_file_name_mm4(input_str)
    print(f"MM4 Hash: {mm4}")
    print(f"MM4 Hex : {mm4 & 0xFFFF:04x}")

    mm3 = hash_file_name_mm3(input_str)
    print(f"MM3 Hash: {mm3}")
    print(f"MM3 Hex : {mm3 & 0xFFFF:04x}")

if __name__ == "__main__":
    main()